pennylane/interfaces/autograd.py

```diff
@@ -14,47 +14,83 @@
+            print(f"Calculating the gradient with {x}, {dy}")
+            return dy * exponent * x**(exponent-1)
+        return grad_fn
+
+    autograd.extend.defvjp(f, vjp, argnums=[0])
+
+
+>>> autograd.grad(f)(autograd.numpy.array(2.0))
+Calculating the gradient with 2.0, 1.0
+4.0
+
+The above code told autograd how to differentiate the first argument of ``f``.
+
+We have an additional problem that autograd does not understand that a :class:`~.QuantumTape`
+contains parameters we want to differentiate. So in order to match the ``vjp`` function with
+the correct parameters, we need to extract them from the batch of tapes, and pass them as is
+as the first argument to the primitive. Even though the primitive function (``f``/ ``_execute``)
+does not use the parameters, that is how we communicate to autograd what parameters the derivatves
+belong to.
+
+
 """
 # pylint: disable=too-many-arguments
 import logging
-import inspect
+from typing import Tuple, Callable
 
 import autograd
 from autograd.numpy.numpy_boxes import ArrayBox
 
 import pennylane as qml
 
+Batch = Tuple[qml.tape.QuantumTape]
+ExecuteFn = Callable[[Batch], qml.typing.ResultBatch]
 
 logger = logging.getLogger(__name__)
 logger.addHandler(logging.NullHandler())
 
 
-def execute(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_diff=2):
+def autograd_execute(
+    tapes: Batch,
+    execute_fn: ExecuteFn,
+    jpc: qml.interfaces.jacobian_products.JacobianProductCalculator,
+):
     """Execute a batch of tapes with Autograd parameters on a device.
 
     Args:
         tapes (Sequence[.QuantumTape]): batch of tapes to execute
-        device (pennylane.Device): Device to use to execute the batch of tapes.
-            If the device does not provide a ``batch_execute`` method,
-            by default the tapes will be executed in serial.
-        execute_fn (callable): The execution function used to execute the tapes
-            during the forward pass. This function must return a tuple ``(results, jacobians)``.
-            If ``jacobians`` is an empty list, then ``gradient_fn`` is used to
-            compute the gradients during the backwards pass.
-        gradient_kwargs (dict): dictionary of keyword arguments to pass when
-            determining the gradients of tapes
-        gradient_fn (callable): the gradient function to use to compute quantum gradients
-        _n (int): a positive integer used to track nesting of derivatives, for example
-            if the nth-order derivative is requested.
-        max_diff (int): If ``gradient_fn`` is a gradient transform, this option specifies
-            the maximum order of derivatives to support. Increasing this value allows
-            for higher order derivatives to be extracted, at the cost of additional
-            (classical) computational overhead during the backwards pass.
+        execute_fn (Callable[[Sequence[.QuantumTape]], ResultBatch]): a function that turns a batch of circuits into results

```

Comment:
TIL how Callable specifies return type in type-hints. I find this extremely unintuitive... but you are definitely following [the docs](https://docs.python.org/3/library/typing.html#annotating-callable-objects) so I guess I ought to drop it. but man, it really looks like the results are arguments.

Reply:
Maybe I could just put `Callable` here and then use words to say "a function from a batch of tapes to results"?

Reply:
I think it's fine as it is. working on picking my battles 😁